### PR TITLE
[WIP][OLD] Adding Stride Diagnostics 

### PR DESCRIPTION
**Pull request by IXLLEGACYIXL**
# PR Details

Adding a Source Generator that Analyzes Projects for Violations of how Stride would expect it to be

## Description

The Analyzer Covers:
1. Access of Collections Properties
2. Access of Array Properties
3. Acces of Normal Properties
4. cant Datamember something that is private or a DataMemberIgnore
5. cant have a DataMember on a Dictionary with a object as key

## Related Issue

n/a

## Motivation and Context

It solves the issue that you get surprised when you open Game Studio and you don't see anything Exported
Also it offers validation for other generators so they dont have to check the same errors

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] My change requires a change to the documentation. ( Updating the docs + making sites for the diagnostics )
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: sources/core/Stride.Core.DiagnosticsServices.Tests/GlobalUsings.cs [hunk 1]
```diff
@@ -0,0 +1 @@
+global using Xunit;
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices.Tests/GlobalUsings.cs:1]: Is this file needed if you're including `using Xunit` everywhere?

File: sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="coverlet.collector" Version="3.2.0">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\Stride.Core.DiagnosticsServices\Stride.Core.DiagnosticsServices.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="true" />
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj:25]: Is `OutputItemType="Analyzer"` valid on the reference in the test project - isn't it making the analyzer run over the test project as well?

File: sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    <PackageReference Include="Stride.Engine" Version="4.1.0.1838" />
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj:26]: We can't use PackageReference for projects in the same solution. Use ProjectReference instead. But also, can you only include `Stride.Core` instead of the whole Engine?

File: sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.4">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.4.0" />
+    <PackageReference Include="xunit" Version="2.5.0" />
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj:34]: Duplicate `xunit` reference

File: sources/core/Stride.Core.DiagnosticsServices/ClassInfo.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
+﻿using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis;
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices/ClassInfo.cs:2]: Please sort using in the files. VS > Edit > Intellisense > Sort usings

File: sources/core/Stride.Core.DiagnosticsServices/ClassInfo.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+                IEnumerable<IPropertySymbol> properties = PropertyAttributeFinder.FilterBasePropertiesRecursive(ref classSymbol);
+                return _propertyCache = properties.ToList();
+            }
+            else
+            {
+                return _propertyCache;
+            }
+        }
+    }
+    public bool IsAbstract()
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices/ClassInfo.cs:41]: Should this be a property as well, like `IsGeneric`?

File: sources/core/Stride.Core.DiagnosticsServices/Diagnoser.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+internal class Diagnoser
+{
+    internal void StartCreation(ClassInfo info)
+    {
+        DiagnoseDataMember(info);
+    }
+    private void DiagnoseDataMember(ClassInfo info)
+    {
+        IEnumerable<Type> reporterTypes = typeof(Diagnoser).Assembly.GetTypes()
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices/Diagnoser.cs:17]: This could be cached by assigning it to a static field
```csharp
private static List<Type> reporterTypes = typeof(Diagnoser).Assembly.GetTypes()
    .Where(type => typeof(IViolationReporter).IsAssignableFrom(type) && !type.IsInterface && !type.IsAbstract)
    .ToList();
```

File: sources/core/Stride.Core.DiagnosticsServices/ErrorCodes.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace StrideDiagnostics;
+public static class ErrorCodes
+{
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices/ErrorCodes.cs:7]: Would be good to add XML comments to each of those constant with some explanation when it is raised.

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+namespace StrideDiagnostics;
+internal class PropertyAttributeFinder
+{
+    /// <summary>
+    /// Walks through a base class of a class and retrieves all allowed Properties.
+    /// Then it tries to get it's own base class and get from it all allowed Properties recursively.
+    /// All the Properties get summed up.
+    /// If they fulfill :
+    ///     Proper Accessors <see cref="PropertyHasAllowedAccessors(IPropertySymbol)"/>
+    ///     Shouldnt be Ignored <see cref="HasDataMemberIgnoreAttribute(IPropertySymbol)"/>
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:21]: Is `HasDataMemberIgnoreAttribute` method missing?

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        return result;
+    }
+
+    private static bool PropertyHasAllowedAccessors(IPropertySymbol propertyInfo)
+    {
+        if (propertyInfo == null)
+            return false;
+        return (propertyInfo.SetMethod?.DeclaredAccessibility == Accessibility.Public ||
+                propertyInfo.SetMethod?.DeclaredAccessibility == Accessibility.Internal ||
+                propertyInfo.GetMethod?.ReturnsVoid == true
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:42]: When does a getter return void? Did you mean `SetMethod`?

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        List<IPropertySymbol> result = new List<IPropertySymbol>();
+        while (currentBaseType != null)
+        {
+            result.AddRange(currentBaseType.GetMembers().OfType<IPropertySymbol>().Where(PropertyHasAllowedAccessors));
+            currentBaseType = currentBaseType.BaseType;
+        }
+        return result;
+    }
+
+    private static bool PropertyHasAllowedAccessors(IPropertySymbol propertyInfo)
```
Comment (manio143) [inline sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:36]: Note that Stride serialization supports get-only properties for collections and classes where the contents are assigned to the instance. There is more complicated logic needed here. See `Stride.Core.DataMemberMode`

**Review by manio143 — COMMENTED** [review]
Part 1 of my review - It's a good start and I like the `IViolationReporter` approach.
I'll have another go at it tomorrow.

File: sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        // Assert that there is an error
+        Assert.True(!hasError, "The Array should be ignored but wasn't.");
+
+    }
+    [Fact]
+    public void IgnoreMember2()
+    {
+        // Define the source code for the Class1 class with an invalid property
+        string sourceCode = @"
+using System.Runtime.Serialization;
```
[the remark this answers is not in the document]
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs:58]: i will refactor it, its a left over when i had a bug that DataMember of System was checked and Strides..
i compare now also against the dll in which the attribute is

This is not yet true for DataContract, there it doesnt matter yet

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs (hunk 9 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:21]: Yes i moved it to the PropertyAttributeFinderExtensions, will remove it

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs (hunk 10 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:42]: i think it returns void if its not existent , i will check it out

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs (hunk 11 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:36]: currently get only properties work with the check

the assigning isnt covered yet, i gotta check it out

File: sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
+﻿using Microsoft.CodeAnalysis;
+using System.Runtime.Serialization;
+using Xunit;
+
+namespace StrideDiagnosticsTests;
```
[the remark this answers is not in the document]
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs:5]: https://github.com/stride3d/stride/pull/1778/commits/e60c88e1142bf9bcfae59664df987acffab16ba7

File: sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+using System.Runtime.Serialization;
```
[the remark this answers is not in the document]
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs:6]: [b1de034](https://github.com/stride3d/stride/pull/1778/commits/b1de03408bdafa7127f2694bc8c81c3f02b0580a)

there was no intention for having different behaviour, adjusted all of them.

File: sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+public class ArrayError
```
[the remark this answers is not in the document]
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs:8]: [6a388bd](https://github.com/stride3d/stride/pull/1778/commits/6a388bde8474bae1a689f388d4376d70c6bb364a)
probably they still need better naming but it should be better now

File: sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs:58]: [4b6fb3b](https://github.com/stride3d/stride/pull/1778/commits/4b6fb3b46a5ce422eca1d5e4797ce6f6a35f07e0)
updated all of them which are remaining

File: sources/core/Stride.Core.DiagnosticsServices.Tests/GlobalUsings.cs (hunk 1 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/GlobalUsings.cs:1]: ive changed it to use csproj usings and removed it everywhere in the cleanup

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs (hunk 9 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:21]: https://github.com/stride3d/stride/pull/1778/commits/7e5ef6e33c2ae8223e8e5e7258b4ce49b8ef389b

was removed as its unnecessary anyway

File: sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+{
+    [DataMemberIgnore]
+    public ArrayError[] Array { get; set; }
+}";
+        IEnumerable<Diagnostic> generatedDiagnostics = DiagnosticsHelper.GetDiagnostics(sourceCode);
+        // Check if there are any diagnostics with the expected ID
+        bool hasError = generatedDiagnostics.Any();
+
+        // Assert that there is an error
+        Assert.True(!hasError, "The Array should be ignored but wasn't.");
```
[the remark this answers is not in the document]
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/ArrayError.cs:50]: [d43de05](https://github.com/stride3d/stride/pull/1778/commits/d43de0589db93b9e7a3e8c2e153f194c772a6fc7)

replaced all remaining assert true

File: sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj (hunk 2 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj:25]: f2b8bfa
removed it , was an leftover from my tries

File: sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj (hunk 3 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj:26]: 3a302bd

was replaced to a projectreference

File: sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj (hunk 4 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices.Tests/Stride.Core.DiagnosticsServices.Tests.csproj:34]: was removed by the latest adding of the test project, doesnt exist anymore

File: sources/core/Stride.Core.DiagnosticsServices/ClassInfo.cs (hunk 6 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/ClassInfo.cs:41]: [29d7e7d](https://github.com/stride3d/stride/pull/1778/commits/29d7e7d2b76ce6c56cf0063307a2aee8b10bf015)

is now a property

File: sources/core/Stride.Core.DiagnosticsServices/Diagnoser.cs (hunk 7 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/Diagnoser.cs:17]: c0592db

is applied

File: sources/core/Stride.Core.DiagnosticsServices/ErrorCodes.cs (hunk 8 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/ErrorCodes.cs:7]: [b2c044c](https://github.com/stride3d/stride/pull/1778/commits/b2c044c34d1961474eaad75ce37169fe8b8e0875)

ive added base comments and renamed them, the documentation is overall really bad for now :D

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs (hunk 10 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:42]: [f7d15a0](https://github.com/stride3d/stride/pull/1778/commits/f7d15a0e53ee996ea347365f8bd9f13b8b91296a)
it will never be true for normal getters

File: sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs (hunk 11 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.DiagnosticsServices/PropertyAttributeFinder.cs:36]: could you give me an example when my current check would fail?

**Comment by IXLLEGACYIXL** [comment]
i will rewrite the enumerable logic as im iterating to much 
the inclusion of the generator still only partially works and i dont know how to fix

**Comment by IXLLEGACYIXL** [comment]
not needed anymore
